Implement poor man's invitation via Organization invitation

mprasil opened the pull request:
This should allow invitation-only registration via Organization invitations. The way it works follows the suggested approach [outlined](https://github.com/dani-garcia/bitwarden_rs/issues/40#issue-331759674) in #40. When Admin/Owner invites user with an email that is not found, a new record in `invitations` table is created. At the same time an `User` is created with random password and key generated just as a "placeholder".

When user tries to register, they will be allowed to do so even if SIGNUPS_ALLOWED is set to false if their email is in the invitations table.

This should also allow user to register without the placeholder `User` already being present, which would accommodate some external integration like the one [suggested](https://github.com/dani-garcia/bitwarden_rs/issues/40#issuecomment-405399443) by @ViViDboarder. But we'd have to document that kind of usage.

dani-garcia commented:
Is there a reason for generating a random password and hashing it instead of just setting the password_hash to an empty vec directly? That way there is no need for a new dependency.

Also, all the Invitation::find_by_mail are followed by a delete, so maybe we can create a method take_by_mail or something that returns the invitation if there is one and also deletes it. It would remove some matches and make the code easier to follow. (Maybe we can just use the diesel::delete function in that hypothetical take_by_mail, which I think returns the number of rows deleted, as a check that the invitation existed).

mprasil commented:
The reason mostly is, that User::new() will take that empty vec and will hash it to save as password hash, so this would in theory add valid user with blank password, that could actually log in. I'm not sure if that's bad thing. I can definitely remove that part if you think that makes sense.

There is one exception to that delete rule: When the invitation is present, but the `INVITATIONS_ALLOWED` are set to false. I didn't want to delete the invitation in case the invitations are allowed later on.

dani-garcia commented:
Well, if you use the User::new method as it is, then yeah anyone sending a blank masterPasswordHash could log in, because that would produce a valid encrypted hash. But if you directly set the password_hash vec to empty, bypassing that method, then the crypto::verify_password_hash method would fail (we could even add a check in there to make sure, but no input should produce and empty vec hash).

About the invitations thing, that could be solved by checking the INVITATIONS_ALLOWED flag first, that way we also do one less DB call.

mprasil commented:
Thanks a lot for your input. I've rewrote a lot of the implementation to hopefully make it nicer. I've also added documentation and I've added code to make invited, but unregistered users handled as "Invited" and not "Accepted" which makes it work a bit better in the Org admin interface.

I've squashed all the commits into one as it was already a mess.

dani-garcia commented:
I think it does look nicer now.

The user::new_invited won't work correctly like that, that empty string would get hashed correctly here:
https://github.com/dani-garcia/bitwarden_rs/blob/master/src/db/models/user.rs#L48.

As it is anyone sending an empty masterPasswordHash could login in that account.

I think the best solution would be to remove the password parameter from the user::new function and set the password_hash to Vec::new(), then in the cases where we do set the password, we would explicitly call the already existing set_password function (Which I think it's only done in the register method).

mprasil commented:
Oh, my bad! I forgot to handle the hash manually as you suggested while working on the other changes.

mprasil commented:
OKay I've did the changes. We now need to set the key and password in extra steps, the `User::new()` will set those to blank.

dani-garcia commented:
Okay, looking great now, thanks!
